timemachine/cpp/src/kernels/k_neighborlist.cuh

```diff
@@ -22,71 +21,92 @@ void __global__ k_find_block_bounds(
-    RealType pos_z = coords[atom_idx * 3 + 2];
-
-    RealType minPos_x = pos_x;
-    RealType minPos_y = pos_y;
-    RealType minPos_z = pos_z;
-
-    RealType maxPos_x = pos_x;
-    RealType maxPos_y = pos_y;
-    RealType maxPos_z = pos_z;
-
-    const int last = min(row_idx + TILE_SIZE, num_indices);
-    for (int i = row_idx + 1; i < last; i++) {
-        atom_idx = row_idxs[i];
+
+    if (row_idx < num_indices) {
+        int atom_idx = row_idxs[row_idx];
+
         pos_x = coords[atom_idx * 3 + 0];
         pos_y = coords[atom_idx * 3 + 1];
         pos_z = coords[atom_idx * 3 + 2];
-        // Build up center over time, and recenter before computing
-        // min and max, to reduce overall size of box thanks to accounting
-        // for periodic boundary conditions
-        RealType center_x = static_cast<RealType>(0.5) * (maxPos_x + minPos_x);
-        RealType center_y = static_cast<RealType>(0.5) * (maxPos_y + minPos_y);
-        RealType center_z = static_cast<RealType>(0.5) * (maxPos_z + minPos_z);
-        pos_x -= bx * nearbyint((pos_x - center_x) * inv_bx);
-        pos_y -= by * nearbyint((pos_y - center_y) * inv_by);
-        pos_z -= bz * nearbyint((pos_z - center_z) * inv_bz);
-        minPos_x = min(minPos_x, pos_x);
-        minPos_y = min(minPos_y, pos_y);
-        minPos_z = min(minPos_z, pos_z);
-
-        maxPos_x = max(maxPos_x, pos_x);
-        maxPos_y = max(maxPos_y, pos_y);
-        maxPos_z = max(maxPos_z, pos_z);
-    }
 
-    block_bounds_ctr[tile_idx * 3 + 0] = static_cast<RealType>(0.5) * (maxPos_x + minPos_x);
-    block_bounds_ctr[tile_idx * 3 + 1] = static_cast<RealType>(0.5) * (maxPos_y + minPos_y);
-    block_bounds_ctr[tile_idx * 3 + 2] = static_cast<RealType>(0.5) * (maxPos_z + minPos_z);
+        min_pos_x = pos_x;
+        min_pos_y = pos_y;
+        min_pos_z = pos_z;
 
-    block_bounds_ext[tile_idx * 3 + 0] = static_cast<RealType>(0.5) * (maxPos_x - minPos_x);
-    block_bounds_ext[tile_idx * 3 + 1] = static_cast<RealType>(0.5) * (maxPos_y - minPos_y);
-    block_bounds_ext[tile_idx * 3 + 2] = static_cast<RealType>(0.5) * (maxPos_z - minPos_z);
+        max_pos_x = min_pos_x;
+        max_pos_y = min_pos_y;
+        max_pos_z = min_pos_z;
+    }
 
-    // Reset the ixn count
-    if (tile_idx == 0) {
-        ixn_count[0] = 0;
+    // Build up center over time, and recenter before computing
+    // min and max, to reduce overall size of box thanks to accounting
+    // for periodic boundary conditions
+    const int src_lane = threadIdx.x + 1 % WARP_SIZE;
+    for (int i = 0; i < WARP_SIZE; i++) {
+        row_idx = __shfl_sync(0xffffffff, row_idx, src_lane);
+        pos_x = __shfl_sync(0xffffffff, pos_x, src_lane);
+        pos_y = __shfl_sync(0xffffffff, pos_y, src_lane);
+        pos_z = __shfl_sync(0xffffffff, pos_z, src_lane);
+        // Only evaluate for the first thread and when the row idx is valid
+        if (threadIdx.x % WARP_SIZE == 0 && row_idx < num_indices) {
```

Comment:
I'm not 100% clear on what's going on here, but it looks like the `threadIdx.x % WARP_SIZE == 0` check might be moved outside of the for loop?

Reply:
Moved out of the loop in https://github.com/proteneer/timemachine/pull/1105/commits/6e40cad6a5c94ffff4363f6c47d30946f8badce1